metadata/src/main/java/org/apache/kafka/controller/SnapshotGenerator.java

```diff
@@ -0,0 +1,146 @@
+                      SnapshotWriter writer,
+                      int maxBatchesPerGenerateCall,
+                      ExponentialBackoff exponentialBackoff,
+                      List<Section> sections) {
+        this.log = logContext.logger(SnapshotGenerator.class);
+        this.writer = writer;
+        this.maxBatchesPerGenerateCall = maxBatchesPerGenerateCall;
+        this.exponentialBackoff = exponentialBackoff;
+        this.sections = sections;
+        this.sectionIterator = this.sections.iterator();
+        this.batchIterator = Collections.emptyIterator();
+        this.batch = null;
+        this.section = null;
+        this.numRecords = 0;
+        this.numWriteTries = 0;
+    }
+
+    /**
+     * Returns the epoch of the snapshot that we are generating.
+     */
+    long epoch() {
+        return writer.epoch();
+    }
+
+    SnapshotWriter writer() {
+        return writer;
+    }
+
+    /**
+     * Generate the next batch of records.
+     *
+     * @return  0 if a batch was sent to the writer,
+     *          -1 if there are no more batches to generate,
+     *          or the number of times we tried to write and the writer
+     *          was busy, otherwise.
+     */
+    private long generateBatch() throws Exception {
+        if (batch == null) {
+            while (!batchIterator.hasNext()) {
+                if (section != null) {
+                    log.info("Generated {} record(s) for the {} section of snapshot {}.",
+                             numRecords, section.name(), writer.epoch());
+                    section = null;
+                    numRecords = 0;
+                }
+                if (!sectionIterator.hasNext()) {
+                    writer.completeSnapshot();
+                    return -1;
+                }
+                section = sectionIterator.next();
+                log.info("Generating records for the {} section of snapshot {}.",
+                         section.name(), writer.epoch());
+                batchIterator = section.iterator();
+            }
+            batch = batchIterator.next();
+        }
+        if (writer.writeBatch(batch)) {
+            numRecords += batch.size();
+            numWriteTries = 0;
+            batch = null;
+            return 0;
+        } else {
+            return ++numWriteTries;
+        }
+    }
+
+    /**
+     * Generate the next few batches of records.
+     *
+     * @return  The number of nanoseconds to delay before rescheduling the
+     *          generateBatches event, or empty if the snapshot is done.
+     */
+    OptionalLong generateBatches() throws Exception {
+        for (int numBatches = 0; numBatches < maxBatchesPerGenerateCall; numBatches++) {
+            long result = generateBatch();
+            if (result < 0) {
+                return OptionalLong.empty();
+            } else if (result > 0) {
+                return OptionalLong.of(exponentialBackoff.backoff(numWriteTries - 1));
```

Comment:
Should we use `result` here instead of `numWriteTries`? If not, maybe we can change generateBatch() to return a boolean instead

Reply:
It's fine to use `result` here, I guess.